File: subway/taxi.py
"""Taxi / ride-hail: the zone-to-zone flow model (scripts/build_taxi.py) and the
street paths for each zone pair (scripts/build_streets.py).

There is no realtime feed for cabs, so the browser replays a "typical hour" of
trips as particles. Each particle follows an actual driving route between its
pickup and drop-off zones, including going over bridges and through tunnels.
"""
from __future__ import annotations

import json
import struct
import threading
from collections import OrderedDict
from pathlib import Path

try:
    import duckdb
except ImportError:  # pragma: no cover - optional dependency
    duckdb = None

try:
    import numpy as np
except ImportError:  # pragma: no cover
    np = None

SERVICES = ("uber", "lyft", "yellow", "other_hv")
TRIPS_MAGIC = 0x51495854  # "TXIQ"


class StreetPaths:
    """Memory-mapped routed paths for zone pairs."""

    def __init__(self, streets_dir: Path):
        self.dir = Path(streets_dir)
        self.index: dict[str, list[list[int]]] = {}
        self.verts = None
        idx, npy = self.dir / "paths_index.json", self.dir / "paths.npy"
        if np is not None and idx.exists() and npy.exists():
            self.index = json.loads(idx.read_text())
            self.verts = np.load(npy, mmap_mode="r")

    @property
    def available(self) -> bool:
        return self.verts is not None and bool(self.index)

    def variants(self, pu: int, do: int) -> list[list[int]]:
        return self.index.get(f"{pu}-{do}", [])


class TaxiFlow:
    def __init__(self, db_path: Path, zones_path: Path, streets_dir: Path):
        self.db_path = Path(db_path)
        self.zones_path = Path(zones_path)
        self.paths = StreetPaths(streets_dir)
        self._lock = threading.Lock()
        self._con = None
        self._zones_bytes: bytes | None = None
        self._meta: dict | None = None
        self._cache: OrderedDict[tuple, dict] = OrderedDict()

    @property
    def available(self) -> bool:
        return duckdb is not None and self.db_path.exists() and self.zones_path.exists()

    def _connect(self):
        if self._con is None:
            self._con = duckdb.connect(str(self.db_path), read_only=True)
            rows = self._con.execute("SELECT month, dow, days FROM meta").fetchall()
            self._meta = {"month": rows[0][0] if rows else None, "days": {int(d): int(n) for _, d, n in rows}}
        return self._con

    def status(self) -> dict:
        if not self.available:
            return {"available": False, "hint": "run: python scripts/build_taxi.py (downloads ~600 MB of TLC data)"}
        with self._lock:
            self._connect()
            return {
                "available": True,
                "month": self._meta["month"],
                "services": list(SERVICES),
                "realtime": False,
                "street_paths": self.paths.available,
            }

    def zones_bytes(self) -> bytes:
        if self._zones_bytes is None:
            self._zones_bytes = self.zones_path.read_bytes() if self.zones_path.exists() else b'{"zones":[]}'
        return self._zones_bytes

    # ---- flows ------------------------------------------------------------------
    def flow(self, dow: int, hour: int, services: list[str], top: int = 4000) -> dict:
        """Per-day-average flows for a weekday (0=Mon) and hour, for the chosen services."""
        if not self.available:
            return {"available": False, "hint": "run: python scripts/build_taxi.py", "flows": [], "zones": []}
        dow, hour = dow % 7, hour % 24
        services = sorted(s for s in services if s in SERVICES) or list(SERVICES)
        key = (dow, hour, tuple(services), top)
        with self._lock:
            cached = self._cache.get(key)
            if cached is not None:
                self._cache.move_to_end(key)
                return cached
            result = self._query(dow, hour, services, top)
            self._cache[key] = result
            while len(self._cache) > 12:
                self._cache.popitem(last=False)
            return result

    def _rows(self, dow: int, hour: int, services: list[str]) -> list[tuple]:
        """Every drawable flow row for that hour: (pu, dropoff, trips, avg_secs, service)."""
        con = self._connect()
        placeholders = ",".join("?" * len(services))
        return con.execute(
            f"""
            SELECT pu, dropoff, trips, avg_secs, service
            FROM flows WHERE dow = ? AND hour = ? AND service IN ({placeholders})
              AND pu BETWEEN 1 AND 263 AND dropoff BETWEEN 1 AND 263 AND pu <> dropoff
            ORDER BY trips DESC, pu, dropoff, service
            """,
            [dow, hour, *services],
        ).fetchall()

    def _query(self, dow: int, hour: int, services: list[str], top: int) -> dict:
        con = self._connect()
        placeholders = ",".join("?" * len(services))
        days = max(1, self._meta["days"].get(dow, 4))
        flows = self._rows(dow, hour, services)[:top]
        # trips we cannot draw: unknown zones (264/265) and trips that start and end in one zone
        undrawn = con.execute(
            f"""
            SELECT coalesce(sum(CASE WHEN pu = dropoff THEN trips END), 0)::INTEGER,
                   coalesce(sum(CASE WHEN pu NOT BETWEEN 1 AND 263 OR dropoff NOT BETWEEN 1 AND 263 THEN trips END), 0)::INTEGER
            FROM flows WHERE dow = ? AND hour = ? AND service IN ({placeholders})
            """,
            [dow, hour, *services],
        ).fetchone()
        zones = con.execute(
            f"""
            WITH f AS (SELECT * FROM flows WHERE dow = ? AND hour = ? AND service IN ({placeholders}))
            SELECT z.id,
                   coalesce((SELECT sum(trips) FROM f WHERE f.pu = z.id), 0)::INTEGER AS pickups,
                   coalesce((SELECT sum(trips) FROM f WHERE f.dropoff = z.id), 0)::INTEGER AS dropoffs
            FROM zones z
            """,
            [dow, hour, *services],
        ).fetchall()
        by_service = con.execute(
            f"""
            SELECT service, sum(trips)::INTEGER FROM flows
            WHERE dow = ? AND hour = ? AND service IN ({placeholders}) GROUP BY service
            """,
            [dow, hour, *services],
        ).fetchall()
        total = sum(n for _, n in by_service)
        return {
            "available": True,
            "month": self._meta["month"],
            "dow": dow,
            "hour": hour,
            "days_in_sample": days,
            "services": services,
            "street_paths": self.paths.available,
            # everything below is divided by the number of such weekdays in the
            # month, i.e. it's the average for ONE such hour.
            "trips_per_hour": round(total / days),
            "by_service": {s: round(n / days) for s, n in by_service},
            # trips in the hour that the particles cannot show
            "same_zone_per_hour": round(undrawn[0] / days),
            "unknown_zone_per_hour": round(undrawn[1] / days),
            "drawable_per_hour": round((total - undrawn[0] - undrawn[1]) / days),
            # [pickup_zone, dropoff_zone, trips_per_hour, avg_trip_seconds, service, n_path_variants]  (busiest pairs)
            "flows": [
                [pu, do, round(n / days, 2), secs, svc, len(self.paths.variants(pu, do))] for pu, do, n, secs, svc in flows
            ],
            "zones": [[zid, round(p / days, 1), round(d / days, 1)] for zid, p, d in zones],
        }

    # ---- sampled trips with their street paths -------------------------------------
    def trips_bytes(self, dow: int, hour: int, services: list[str], n: int) -> bytes:
        """`n` trips drawn from the WHOLE zone-to-zone distribution for that hour.

        Systematic sampling (evenly spaced points along the cumulative trip count, with
        a seeded offset) so every pair gets particles in proportion to its trips - a pair
        with 0.3 trips/hour gets a particle 30 % of the time - and the same hour always
        looks the same. Each particle stands for `tripsPerDot` real trips per hour.

        Layout (little-endian):
          uint32 magic 'TXIQ', uint32 nTrips, uint32 nPaths, uint32 nVerts,
          float32 tripsPerDot, uint32 daysInSample
          uint32 pathVertStart[nPaths + 1]        vertices of path j are [start[j], start[j+1])
          {uint16 pu, uint16 dropoff, uint16 service, uint16 avg_secs, int32 path} [nTrips]
          float32 verts[nVerts * 3]               lon, lat, z (metres relative to street level)
        `service` indexes SERVICES; `path` is -1 when the pair has no routed street path.
        """
        rows = self._rows(dow, hour, services) if self.available else []
        days = max(1, self._meta["days"].get(dow, 4)) if rows else 1
        n = max(1, min(int(n), 80000))
        if not rows:
            return struct.pack("<IIIIfI", TRIPS_MAGIC, 0, 0, 0, 0.0, days) + struct.pack("<I", 0)

        trips = np.array([r[2] for r in rows], dtype=np.float64)
        cum = np.cumsum(trips)
        total = float(cum[-1])
        step = total / n
        rng = np.random.default_rng(dow * 24 + hour)
        points = rng.random() * step + step * np.arange(n)
        idx = np.minimum(np.searchsorted(cum, points, side="right"), len(rows) - 1)
        counts = np.bincount(idx, minlength=len(rows))

        svc_index = {s: i for i, s in enumerate(SERVICES)}
        path_ids: dict[tuple[int, int], int] = {}
        path_chunks: list = []
        path_start = [0]
        n_verts = 0
        recs = np.zeros(n, dtype=[("pu", "<u2"), ("do", "<u2"), ("svc", "<u2"), ("secs", "<u2"), ("path", "<i4")])
        k = 0
        for i in np.nonzero(counts)[0]:
            pu, do, _, secs, svc = rows[i]
            variants = self.paths.variants(pu, do) if self.paths.available else []
            for j in range(int(counts[i])):
                pid = -1
                if variants:
                    off, ln = variants[j % len(variants)]
                    key = (off, ln)
                    pid = path_ids.get(key)
                    if pid is None:
                        pid = len(path_start) - 1
                        path_ids[key] = pid
                        path_chunks.append(self.paths.verts[off : off + ln])
                        n_verts += ln
                        path_start.append(n_verts)
                recs[k] = (pu, do, svc_index.get(svc, 3), min(int(secs or 0), 65535), pid)
                k += 1
        recs = recs[:k]
        verts = np.concatenate(path_chunks) if path_chunks else np.zeros((0, 3), np.float32)
        head = struct.pack("<IIIIfI", TRIPS_MAGIC, len(recs), len(path_start) - 1, n_verts, step / days, days)
        return (
            head
            + np.asarray(path_start, dtype=np.uint32).tobytes()
            + recs.tobytes()
            + np.ascontiguousarray(verts, dtype=np.float32).tobytes()
        )

File: web/js/taxi.js
// Taxi / Uber / Lyft: replay a typical hour of trips as particles that drive
// real street routes (over bridges, through tunnels) between taxi zones.
//
// The server samples `particles` trips from the whole zone-to-zone distribution
// for the chosen weekday + hour (see subway/taxi.py), so each dot stands for the
// same number of real trips and every neighbourhood gets its true share. Paths
// are expanded here into flat typed arrays for deck.gl's binary attribute mode.

import { decodeTaxiTrips, getBuffer, getJSON } from "./api.js";
import { M_PER_DEG_LAT, M_PER_DEG_LON, clamp } from "./geo.js";

export const LOOP_SECONDS = 3600; // one hour of trips ...
export const SERVICES = ["uber", "lyft", "yellow", "other_hv"];
export const SERVICE_COLORS = { uber: [235, 235, 235], lyft: [255, 0, 191], yellow: [255, 212, 0], other_hv: [120, 200, 255] };
export const DENSITIES = { sparse: 7000, medium: 20000, full: 0 }; // 0 = one dot per trip

export class TaxiReplay {
  constructor(opts = {}) {
    this.particleTarget = opts.particles ?? 7000; // 0 => one dot per drawable trip
    this.zones = null; // id -> zone
    this.zoneList = [];
    this.flow = null;
    this.trips = []; // metadata per particle (for tooltips)
    this.tripsPerDot = 0;
    this.data = null; // binary TripsLayer data
    this.pathsAvailable = false;
    this._req = 0;
  }

  get count() { return this.trips.length; }

  async loadZones() {
    const z = await getJSON("/api/taxi/zones");
    this.zones = {};
    for (const zone of z.zones) this.zones[zone.id] = zone;
    this.zoneList = z.zones;
  }

  /** Load flow summary + sampled trips for a weekday/hour/service selection. Resolves to the flow summary. */
  async load(dow, hour, services) {
    const req = ++this._req;
    const q = `dow=${dow}&hour=${hour}&services=${[...services].join(",")}`;
    const flow = await getJSON(`/api/taxi/flow?${q}&top=200`);
    if (req !== this._req) return null;
    const n = this.particleTarget || Math.max(1, flow.drawable_per_hour || flow.trips_per_hour);
    const bundle = await getBuffer(`/api/taxi/trips?${q}&n=${n}`).then(decodeTaxiTrips).catch(() => null);
    if (req !== this._req) return null; // superseded
    this.flow = flow;
    this.pathsAvailable = !!flow.street_paths;
    this.build(bundle);
    return flow;
  }

  jitter(zone) {
    // spread pickups across the zone: ~ ±350 m gaussian-ish
    const r = (Math.random() + Math.random() + Math.random() - 1.5) * 0.006;
    const a = Math.random() * Math.PI * 2;
    return [zone.lon + Math.cos(a) * r, zone.lat + Math.sin(a) * r * 0.76, 0];
  }

  build(bundle) {
    this.trips = [];
    this.data = null;
    this.tripsPerDot = 0;
    if (!bundle || !bundle.trips.length || !this.zones) return;
    const { trips, pathStart, verts } = bundle;
    this.tripsPerDot = bundle.tripsPerDot;

    // first pass: how many particles and vertices (trips that spill past the hour are drawn twice)
    const starts = new Float64Array(trips.length);
    const spill = new Uint8Array(trips.length);
    let nParticles = 0, nVerts = 0;
    for (let i = 0; i < trips.length; i++) {
      const t = trips[i];
      const m = t.path >= 0 ? pathStart[t.path + 1] - pathStart[t.path] : 2;
      starts[i] = Math.random() * LOOP_SECONDS;
      const dur = Math.max(120, Math.min(LOOP_SECONDS, t.secs));
      spill[i] = starts[i] + dur > LOOP_SECONDS ? 1 : 0;
      const copies = 1 + spill[i];
      nParticles += copies;
      nVerts += copies * m;
    }

    const positions = new Float32Array(nVerts * 3);
    const timestamps = new Float32Array(nVerts);
    const colors = new Uint8Array(nVerts * 3);
    const startIndices = new Uint32Array(nParticles + 1);
    const meta = new Array(nParticles);
    let p = 0, v = 0;
    const scratch = new Float64Array(4096);

    for (let i = 0; i < trips.length; i++) {
      const t = trips[i];
      const color = SERVICE_COLORS[SERVICES[t.service]] || SERVICE_COLORS.other_hv;
      const start = starts[i];
      let m, base, dur, ox = 0, oy = 0, a = null, b = null;
      if (t.path >= 0) {
        base = pathStart[t.path] * 3;
        m = pathStart[t.path + 1] - pathStart[t.path];
        // lateral offset of a few metres so particles on the same route form lanes
        const side = (Math.random() - 0.5) * 9;
        ox = side / M_PER_DEG_LON; oy = side / M_PER_DEG_LAT;
        // cumulative length along the path -> constant speed timestamps
        const cum = m <= scratch.length ? scratch : new Float64Array(m);
        cum[0] = 0;
        for (let j = 1; j < m; j++) {
          const dx = (verts[base + j * 3] - verts[base + (j - 1) * 3]) * M_PER_DEG_LON;
          const dy = (verts[base + j * 3 + 1] - verts[base + (j - 1) * 3 + 1]) * M_PER_DEG_LAT;
          cum[j] = cum[j - 1] + Math.hypot(dx, dy);
        }
        const length = cum[m - 1] || 1;
        // duration: the observed average for this pair, but keep the implied speed sane
        dur = clamp(t.secs, length / 22, length / 3.5);
        for (let copy = 0; copy <= spill[i]; copy++) {
          const shift = copy ? -LOOP_SECONDS : 0;
          startIndices[p] = v;
          meta[p] = t;
          for (let j = 0; j < m; j++, v++) {
            positions[v * 3] = verts[base + j * 3] + ox;
            positions[v * 3 + 1] = verts[base + j * 3 + 1] + oy;
            positions[v * 3 + 2] = verts[base + j * 3 + 2];
            timestamps[v] = start + shift + (cum[j] / length) * dur;
            colors[v * 3] = color[0]; colors[v * 3 + 1] = color[1]; colors[v * 3 + 2] = color[2];
          }
          p++;
        }
      } else {
        // no routed street path for this pair (islands, Newark ...): fly zone to zone
        const zPu = this.zones[t.pu], zDo = this.zones[t.do];
        if (!zPu || !zDo) continue;
        a = this.jitter(zPu); b = this.jitter(zDo);
        dur = Math.max(120, Math.min(LOOP_SECONDS, t.secs));
        for (let copy = 0; copy <= spill[i]; copy++) {
          const shift = copy ? -LOOP_SECONDS : 0;
          startIndices[p] = v;
          meta[p] = t;
          positions.set(a, v * 3); timestamps[v] = start + shift; colors.set(color, v * 3); v++;
          positions.set(b, v * 3); timestamps[v] = start + shift + dur; colors.set(color, v * 3); v++;
          p++;
        }
      }
    }
    startIndices[p] = v;
    this.trips = meta.slice(0, p);
    this.data = {
      length: p,
      startIndices: startIndices.subarray(0, p + 1),
      attributes: {
        getPath: { value: positions.subarray(0, v * 3), size: 3 },
        getTimestamps: { value: timestamps.subarray(0, v), size: 1 },
        getColor: { value: colors.subarray(0, v * 3), size: 3 },
      },
    };
  }

  /** GeoJSON features for the pickup-volume columns. */
  zoneColumns() {
    if (!this.flow || !this.zoneList.length) return [];
    const m = new Map(this.flow.zones.map(([id, p, d]) => [id, [p, d]]));
    return this.zoneList
      .filter((z) => (m.get(z.id) || [0])[0] > 0)
      .map((z) => ({ type: "Feature", geometry: z.geometry, properties: { id: z.id, zone: z.zone, borough: z.borough, pickups: m.get(z.id)[0], dropoffs: m.get(z.id)[1] } }));
  }
}
